### ISSUE TITLE
RFC: Support multiple descriptions per control

### ISSUE THREAD

**Issue by chris-rock** [inspec/inspec#1695]
### Description

InSpec has already a great support for tags, references and description of controls.  This works great and looks like:

```
control "V-72841" do
  title "PostgreSQL must be configured to prohibit or restrict the use of
        organization-defined functions, ports, protocols, and/or services, as
        defined in the PPSM CAL and vulnerability assessments."
  desc  "In order to prevent unauthorized connection of devices, unauthorized
        transfer of information, or unauthorized tunneling (i.e., embedding of
       ...
       "
```

Sometimes you need additional description e.g. for a specific check or remediation. Right now, those blocks need to be part of the `desc` block, since tags are not suited for long text.

```
  desc "Check": "As the database administrator, run the following SQL:
  $ psql -c \"SHOW port\"
  If the currently defined port configuration is deemed prohibited, this is a
  finding."

  desc "Fix": "Note: The following instructions use the PGDATA environment
  variable. See supplementary content APPENDIX-F for instructions on configuring
  PGDATA.
```

I propose we extend the current `desc` for `control` to allow

```
desc "Section", "Long text"
```

You can still set the default desc by only passing one parameter:
```
control 'my control' do
  desc "A brief description of the control"
  desc "Section", "Long text"

  describe ...
end
```

When overriding a control from a parent the parent description fields will overwrite the child's:
Child control:
 
```
control 'ssh-01' do
  desc "Test to make sure admin password is not admin"
  desc "reasoning", "If this does not work, bad things happen"

  describe ...
end
```

Parent override:

```
include_controls 'child_profile' do
  control 'ssh-01' do
    desc "reasoning", "If this does not work, pull the fire alarm"
    desc "fix", "change the admin password!"
  end
end
```

you should end up with:
```
  desc "Test to make sure admin password is not admin"
  desc  "reasoning", "If this does not work, pull the fire alarm"
  desc "fix", "change the admin password!"
```

For output we will want to add a section to the json schema:
```
"desc:" : "the default description"
^ will stay the same and just be the default description set

"descriptions": [
    { "type": "reason", "data": "why ..." },
^ is a new fields and will be an ARRAY of descriptions
```

To this end we will also need to add descriptions to the inspec/schema fields.

### InSpec and Platform Version

1.20.0

**Comment by arlimus** [comment]
If we establish this as a hash, could this be chained? (just in theory to be aware, not really as a style recommendation)
```
desc "Field1": "val...",
     "Field2": "val..."
```

Also we have to be aware how this behaves in overlays. One idea: overlaying `desc` to change the description will only change the uncategorized description, overlaying `desc "field":...` will change the tagged description.

**Comment by arlimus** [comment]
It will also require additional fields in the JSON structure; hopefully without forcing a major version change.

This would also help us to prevent forcing people from adding descriptions to tagged fields, where we may run into UI problems and which affect inspec query by tag (since we mix tags and descriptions)

**Comment by chris-rock** [comment]
Should we also add warning to inspec check if tags reached a specific length?

**Comment by arlimus** [comment]
That'd be a great idea too, if we take this direction!
Ditto on linebreaks

**Comment by aaronlippold** [comment]
Please also consider adding the ability to line wrap the desc blocks and to 'print pretty' - such as : 

```
desc "check": "As the database administrator, run the following SQL:

  $ psql -c \"SHOW port\"

  If the currently defined port configuration is deemed prohibited, this is a
  finding."

  desc "fix": "Note: The following instructions use the PGDATA environment
  variable. See supplementary content APPENDIX-F for instructions on configuring
  PGDATA.

  To change the listening port of the database, as the database administrator,
  change the following setting in postgresql.conf:

  $ sudo su - postgres
  $ vi $PGDATA/postgresql.conf

  Change the port parameter to the desired port.

  Next, restart the database:

  $ sudo su - postgres
  # SYSTEMD SERVER ONLY
  $ systemctl restart postgresql-9.5
  # INITD SERVER ONLY
  $ service postgresql-9.5 restart

  desc "note": "psql uses the default port 5432 by default. This can be changed by
  specifying the port with psql or by setting the PGPORT environment variable:

  $ psql -p 5432 -c \"SHOW port\"
  $ export PGPORT=5432"
```
To ensure that the human readability is maintained for both the security engineer, the security auditor and the validator. 600 char long strings are hard to read with a \n\n in them. :)

**Comment by aaronlippold** [comment]
you man we shouldn't have:

tag "iloverandomtagsbutthisonewilldo" for a tag name? 🤣 

What would be the general limit for the tag length? 10 chars 20 chars, 40 chars ...

**Comment by chris-rock** [comment]
That would still be okay. It is more the size of the value that concerns us, since this will be difficult to display in UI.

**Comment by aaronlippold** [comment]
Yes, agree. So 32 chars? ( supposing we have someone wanting a "fix_on_aix_or_solaris" or "note_for_time_ sensitive_systems"

**Comment by arlimus** [comment]
@aaronlippold about line-wraps: It is great to enforce this when displaying the profile. Linting should probably also get a warning for extra long lines in description blocks.

Another problem so far has been that line-breaks and "nice formatting" in descriptions have not been respected by the internal engine. This is fixed now, please see https://github.com/chef/inspec/pull/1707 . It now opens the door to e.g. nicely formatted documents, together with the just added fix in https://github.com/chef/inspec/pull/1705 . I'll add another one in the course of today to support indentations for nicer formatting of `InSpec::Control`s

**Comment by aaronlippold** [comment]
So is this the object form we can expect with this RFC?
```
# this is the main control desc - i.e. does not have a "section"
control.add_desc(Inspec::Desc).new(node.xpath('./Rule/description').text.gsub(/\<.*?\>/, '').gsub(/.false/, '.'),wrap:120) 

# control.add_newline ( this may not be needed with the new long-text formatting )
control.add_desc(Inspec::Desc.new('check', node.xpath('./Rule/check/check-content').text),wrap:120)
control.add_desc(Inspec::Desc.new('fix', node.xpath('./Rule/fixtext').text),wrap:120)

# @todo control.ref = 'http://really.important.guadance.pdf'
```

**Comment by arlimus** [comment]
@aaronlippold just a follow-up to my comment [above](https://github.com/chef/inspec/issues/1695#issuecomment-297008507)
Since https://github.com/chef/inspec/pull/1711 is merged too, we now have super pretty `describe` generation 🎊

**Comment by arlimus** [comment]
Merged this comment into: https://github.com/chef/inspec/issues/1695#issuecomment-297939391

**Comment by aaronlippold** [comment]
Sounds good. That structure seems fine to me and what I expected. So just
so I understand, at this point everything's been implemented except the
line wrapping in the to Ruby function.

**Comment by arlimus** [comment]
@aaronlippold Once this RFC is implemented, that's right. The line-wrapper is tricky because we do not make any expectations on the format of the description inside of `desc` calls. To that degree, a line-wrapper may provide some benefits but it also may cause some problems.
However, since you are directly passing things into `Inspec::Control` objects, it should be very feasible to wrap this before passing, where the contents and format of contents is well-known (at least for now).

**Comment by arlimus** [comment]
Aggregation from comments above:

## Solution

* Users can specify multiple description fields with types attached:
    ```
    desc 'deafult ...'
    desc 'reason': 'this is why...'
    desc 'vulnerability': 'it exposes ...'
    ```

* Descriptions are kept as a hash-map internally (not exposed)

* All `desc` fields will have keys that are strings, to prevent unexpected behavior. Users may define keys via
    ```
    desc reason: '...'
    desc 'reason': '...'
    desc 'reason' => '...'
    ```

    which is internally converted into:

    ```
    @desc['reason'] = '...'
    ```

* When working with `Inspec::Control` you access descriptions via:

    ```
    control.desc[''] = 'default desc'
    control.desc['reason'] = '...'
    ```

* Behavior in inheritance: When a control is overwritten, its description fields can be overwritten (or added):
    ```ruby
    control 'base' do
        desc "field": "overwrite..."
        ...
    ```

* A description without any type has the implicit type "default" attached to it:
    ```ruby
    desc "hello..."
    ```

    is equal to

    ```ruby
    desc "default": "hello..."
    ```

    (if we don't go this route, we would have to either leave the type field in the JSON output empty or remove it for the default description, both of which feel shaky)

* The JSON output for description is an array of objects (not a hashmap):
    ```json
    [
        { "type": "...", "data": "description ..." },
    ```

* The ideal desired JSON would have the array of hashes attached to the current keyword `desc`
    ```
    "desc": [ {     }, {     },     ],
    ```

## WIP

The tricky part is the way we will weave this into the JSON output. A couple of options:

1. Add it as a new field. Leave the current desc field untouched and deprecate it + remove in the next major version. It would carry the normal description in duplicate:
    ```json
    "desc": "hello ...",
    "descriptions": [
        { "type": "default", "data": "hello ..." },
        { "type": "reason", "data": "why ..." },
    ```
    Pro: This is non-breaking and has a simple path to deprecation. Con: It blows up JSON contents for now, which would probably put pressure on removing the old field faster.

2. Same as above, just don't put the default description into the `descriptions` array until we bump the major version:
    ```json
    "desc": "hello ...",
    "descriptions": [
        { "type": "reason", "data": "why ..." },
    ```
    It implies that once the major version is bumped and `desc` is removed, its contents will move to `descriptions`. Pro: No data duplication and early access to this feature without bumping the major version. Con: It may require users of the API to adjust their processing if they decide to use this field now, as the default `desc` moves once the major version is bumped.
    
    An alternative to the current wording is to use the fields `desc` and `extra_desc` for now (as string + array) and then once the major version is bumped just replace `desc` with the new behavior.

3. Wait for the major version. Create a separate branch and gate the feature for the next major version branch. It would also allow us to keep the shorthand `desc`.

**Comment by adamleff** [comment]
I feel like I'm going to be the primary dissenter here, but I'm kind of against overloading the `desc` field, though I totally see why users would want this.

Any way I could convince you to consider a new field that only takes keys/values? We could call it `details` or `additional_info`. Added bonus: it would be backward compatible and we wouldn't have to worry about training people to look for a `default` key in a description array.

My gut says this will only be used by a small percentage of InSpec users, even though those InSpec users may have the largest profiles of our entire user base. Given that, I'd advocate for a new field completely.

**Comment by chris-rock** [comment]
@arlimus I am going to prefer `descriptions` in our json output, since it highlights very clear the intention. Also I would prefer to go with option 2, therefore we keep the default description in `desc` and `descriptions` for all additional descriptions.

@adamleff the discussion is only about the internal json representation. Nothing changes for the user: 
```
desc '...'
```
will still work. An alternative is to just go with option 2 and keep the default in `desc` even after version InSpec 2.

Semantically I think `descriptions` is the best term for all this. Essentially we just allow multiple descriptions (structured by sections).  If we add a new word in our json representation, user may not understand the connection between `desc 'reason': '...'` and `additional_info`.

**Comment by arlimus** [comment]
Interesting idea @adamleff . If we were to go with a `details` field, is this what you had in mind?

```ruby
desc "default desc..."
details "reason": "Why we do it this way..."
details "...": "..."
...
```

(I picked `details` from your proposal as it's shorter than `additional_info`; I like it :) )

You probably have a point with many InSpec users probably not actively using it (at least right away). We will probably see a lot of use in the federal space, where this is used extensively. Actually, many profiles will required these fields if they are covered fully in InSpec. For example xccdf supports the `fixtext` and `rationale` fields (used by CIS).

Additionally, there is a long story here with seeing overloaded SCAP descriptions with embedded XML content to covere these fields in STIG (where xccdf was apparently not enough). I'd advocate strongly against introducing keywords for e.g. `rationale`. I think we are on a great track with this idea (both `desc` and `details` can fill this)

Conceptually, I'd love this field to also provide context when controls are overloaded (e.g. criticality changed).

**Comment by arlimus** [comment]
@chris-rock I think Adam meant to suggest a different syntax, see ^^ :)

Thank you for your input on the JSON, that sounds great!!

**Comment by adamleff** [comment]
@arlimus is correct, I'm advocating for a new keyword called `details` to fill this need rather than using `desc` if possible. I think there's value in keeping a single top-level description field present without overloading it to support "additional descriptions" (for lack of a better term). We could keep the existing `desc` field in the JSON just the way it is and introduce a new hash of additional details fields (each with their own key, such as `rationale` as @arlimus gave as an example). 

To be a bit pedantic about it, I see the "description" as a top-level sentence or two about the control itself. If there's additional information about the control that makes it easier for the end-user to reason about it, such as the rationale or the reason or a mitigation process to be used if the control fails, those bits of information don't feel like a "description" of the control but rather additional details the user may find interesting or necessary. I feel they should be broken out accordingly. This has the added benefit of allowing us to rev our JSON schema without introducing any backward compat issues while not reusing keywords. I think seeing "desc" and "descriptions" in a JSON output will be incredibly confusing.

At this point, we all agree that supporting additional arbitrary keyed text fields is a good thing and now just discussing the implementation. If you gents really feel this type of information is actually "description" information, then ignore me. 🙂  But I'd love to see the `desc` field remain as-is given its overall importance.

**Comment by arlimus** [comment]
Just a tiny tiny comment, about the hash of additiona fields in the JSON output: I'd love to stick to
```
[ { key: '...', value: ' ...' }, ... ]
```
over
```
{ 'key...': 'value...', ... }
```
in the JSON. It has to do with integrations and the JSON schema 😉  (I probably just misread it in the comments...)

Also agree that `desc` + `descriptions` in the interim JSON is confusing. We would remove `desc` in that scenario form the JSON after the next major bump, removing this confusion, but I see your point for the interim 👍

**Comment by adamleff** [comment]
@chris-rock and I just chatted, I made my case, and he is going to think about it.

@arlimus I'd actually advocate for:

```
{ 'key...': 'value...', ... }
```

While I know it helps make the schema easier, I think it's harder to work with if I'm looking for specific data. If I was consuming the JSON and wanted to pull out the "rationale" details, it's way easier (and faster) to just grab the "rationale" key from the "details" hash (name TBD!) instead of looping over it until I find the key. Not a HUGE detail, I'm fine either way, but I think the Hash is easier to work with.

**Comment by adamleff** [comment]
@arlimus re: future removal of the `desc` field, I still think that's a bad idea even if it alleviates the technical concern. Having to teach users that an unkeyed "desc" can be found in the "descriptions" "default" key feels like a UX that can be confusing, and I think it dilutes the value of the top-level description of the control.

Thinking from a user's perspective who isn't using keyed descriptions... "Why do I have to grab this 'default' key when I only have one description? I didn't put 'default' anywhere in my control." Is this a valid enough concern I totally just made up? 🙂

**Comment by arlimus** [comment]
@adamleff Just on the Hash vs Array discussion:
I didn't have a strong preference until we released InSpec 1.0. The reason that back then we decided to change from a Hash-model to Arrays has been:

* Schema support
* Integration with platforms (e.g. Elasticsearch) and certain Frameworks

That said I agree that Hashes carry an nice advantage, namely the "faster" access to values (i.e. `fields.key` vs `fields.find('key')`). I personally have this preference too, but I also believe that this representation of our data should be easy to integrate and validate.

In the spirit of consistency I'd like to keep it similar across our output structure. If we want to change it back to Hashes there is a few more fields in the JSON that are affected. Should we open a separate discussion for this?
